### PR TITLE
chore(CPlusPlus): add Matrix Chain Multiplication to Dynamic Programming

### PR DESCRIPTION
**Pull request by ptarafdar001**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] New algorithm
- [ ] Optimization in previous algorithms
- [ ] Code style update
- [ ] Refactor
- [ ] Documentation
- [ ] Other, please describe:

**Briefly describe the changes in this PR**

**Other information:**

### CODE REVIEW

**Comment by ptarafdar001** [comment]
It shows one conflict in **README.md** file, so how can I resolve it?

**Comment by Ashborn-SM** [comment]
@ptarafdar001 i believe you have a merge conflict (not really sure about this).
Did you retrieve changes from the original repo to your remote repo? and did you pull those changes to your local repo before make this commit?

**Comment by ptarafdar001** [comment]
@Ashborn-SM  yeah, I forgot to sync my fork to the original repo. Now, what can I do to fix this issue?

**Comment by Ashborn-SM** [comment]
this could help for syncing https://www.earthdatascience.org/courses/intro-to-earth-data-science/git-github/github-collaboration/update-github-repositories-with-changes-by-others/
And `git pull` to update your local repo.
Although i'm not really sure this will work without a hitch, i'm also not sure this is the solution since i don't know what the actual problem is.

Maybe wait for @Arsenic-ATG, he can help

File: algorithms/CPlusPlus/Dynamic-Programming/matrix-chain-multiplication.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+/*
+ * Time Complexity: O(N3)
+ * Auxiliary Space: O(N2)
+
+ * Input:
+ *  N = 4
+ *  arr = {10, 30, 5, 60}
+ *  Output: 4500
+ * Explaination: The matrices have dimensions
```
Comment (Arsenic-ATG) [inline algorithms/CPlusPlus/Dynamic-Programming/matrix-chain-multiplication.cpp:51]: Typo :
```suggestion
 * Explanation: The matrices have dimensions 
```

File: algorithms/CPlusPlus/Dynamic-Programming/matrix-chain-multiplication.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    for (int i = 0; i < n; i++)
+        cin >> arr[i];
+
+    cout << "Minimum number of multiplications is" << endl;
+    cout << matrixChainMulti(arr, 1, n - 1, dp) << endl;
+}
+
+/*
+ * Time Complexity: O(N3)
+ * Auxiliary Space: O(N2)
```
Comment (Arsenic-ATG) [inline algorithms/CPlusPlus/Dynamic-Programming/matrix-chain-multiplication.cpp:45]:
```suggestion
 * Time Complexity: O(N^3)
 * Auxiliary Space: O(N^2)
```

File: algorithms/CPlusPlus/Dynamic-Programming/matrix-chain-multiplication.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+// Driver Code
+int main()
+{
+    // dp table initialized with -1
+    vector<vector<int>> dp(1001, vector<int>(1001, -1));
+
+    int n;
+    cout << "Enter the array size" << endl;
+    cin >> n;
+    int arr[n];
```
Comment (Arsenic-ATG) [inline algorithms/CPlusPlus/Dynamic-Programming/matrix-chain-multiplication.cpp:34]: Avoid using variable length arrays ( VLAs ) in C++. Although some compilers ( like gcc ) support them as non-standard extension, they are not allowed by C++ standards so a compiler is not obligated to do so.

I see you are already using std::vector in your program then you should have not problem using it here also.

File: algorithms/CPlusPlus/Dynamic-Programming/matrix-chain-multiplication.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// Matrix Chain Multiplicatin - Memorization-Approuch
+
+#include <bits/stdc++.h>
+using namespace std;
+
+// Function for matrix chain multiplication
+int matrixChainMulti(int arr[], int i, int j, vector<vector<int>> dp)
```
Comment (Arsenic-ATG) [inline algorithms/CPlusPlus/Dynamic-Programming/matrix-chain-multiplication.cpp:7]: Is it really mandatory to copy vector `dp` every-time you pass it to the function ?

also I see you not are changing the value of `i` and `j` in the function, then wouldn't it be a good idea to make them const ?

```suggestion
int matrixChainMulti(int arr[],const int i,const int j, vector<vector<int>> &dp)
```

**Review by Arsenic-ATG — CHANGES_REQUESTED** [review]

**Review by Arsenic-ATG — APPROVED** [review]

**Pull request merged**